### PR TITLE
RELATED: RAIL-909 pagable dataSource

### PR DESCRIPTION
**Pull request by tomasreichmann**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/31/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/25/ : SUCCESS in 1m 04s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/32/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/26/ : SUCCESS in 49s

File: src/DataLayer/utils/DummyDataSource.ts [hunk 1]
```diff
@@ -23,6 +76,25 @@ export class DummyDataSource<T> implements IDataSource<T> {
         return Promise.reject('DummyDataSource reject');
     }
 
+    public getPage(
+        resultSpec: AFM.IResultSpec,
+        // tslint:disable-next-line:variable-name
+        _offset: number[],
```
Comment (martinmosko) [inline src/DataLayer/utils/DummyDataSource.ts]: if variable is not used for tslint is enough to write `_: number[]` without using inline comment for disabling tslint check

File: src/execution.ts [hunk 2]
```diff
@@ -14,7 +13,17 @@ import { MetadataModule } from './metadata';
  *
  */
 export class ExecutionModule {
-    constructor(private xhr: XhrModule, private md: MetadataModule) {}
+    public executeAfmModule: ExecuteAfmModule;
```
Comment (martinmosko) [inline src/execution.ts]: `private readonly`

File: src/DataLayer/dataSources/DataSource.ts [hunk 3]
```diff
@@ -1,25 +1,95 @@
...
-import { AFM } from '@gooddata/typings';
+import { AFM, Execution } from '@gooddata/typings';
 import * as stringify from 'json-stable-stringify';
 import { IDataSource } from '../interfaces/DataSource';
 
 export {
     IDataSource
 };
 
+export const SERVER_SIDE_COLUMN_LIMIT = 1000;
```
Comment (martinmosko) [inline src/DataLayer/dataSources/DataSource.ts]: Are you sure its a 1000 and not 500 ?

File: src/DataLayer/dataSources/DataSource.ts [hunk 4]
```diff
@@ -1,25 +1,95 @@
...
+export const SERVER_SIDE_ROW_LIMIT = 1000;
```
Comment (martinmosko) [inline src/DataLayer/dataSources/DataSource.ts]: exporting `SERVER_SIDE_ROW_LIMIT, SERVER_SIDE_COLUMN_LIMIT` for testing does not sound like safe way to test the limitations :) as change does not break any test

please make it private and use separate constant in tests

File: src/DataLayer/dataSources/DataSource.ts (hunk 3 above)
Comment (martinmosko) [reply src/DataLayer/dataSources/DataSource.ts]: https://github.com/gooddata/gooddata-js/blob/master/src/execution/execute-afm.ts#L9 see there is a different limit

File: src/DataLayer/dataSources/DataSource.ts [hunk 5]
```diff
@@ -1,25 +1,95 @@
...
+        [key: string]: Promise<Execution.IExecutionResponse>
+    };
+
     constructor(
         private execFactory: ExecFactory<T>,
         private afm: AFM.IAfm,
-        private fingerprint?: string
-    ) {}
+        private fingerprint?: string,
+        private responseFactory?: (resultSpec: AFM.IResultSpec) => Promise<Execution.IExecutionResponse>,
```
Comment (martinmosko) [inline src/DataLayer/dataSources/DataSource.ts]: This looks to me like something wrong design

I would expect here new `PageableDataSource` inherited from `DataSource`

what is the reason to not doing it this way?

File: src/DataLayer/dataSources/DataSource.ts [hunk 6]
```diff
@@ -1,25 +1,95 @@
...
     }
 
+    public getPage(
+        resultSpec: AFM.IResultSpec,
+        offset: number[] = [],
+        limit: number[] = []
+    ): Promise<Execution.IExecutionResponses> {
+        const resultSpecFingerprint = stringify(resultSpec);
+        if (!this.responseFactory) {
+            throw new Error('Missing responseFactory!');
```
Comment (martinmosko) [inline src/DataLayer/dataSources/DataSource.ts]: this would not be necessary with `PageableDataSource`

File: src/DataLayer/dataSources/DataSource.ts [hunk 7]
```diff
@@ -1,25 +1,95 @@
...
+                    console.warn('The maximum row limit per page is ' + SERVER_SIDE_ROW_LIMIT);
+                }
+                if (columnLimit > safeColumnLimit) {
+                    // tslint:disable-next-line:no-console
+                    console.warn('The maximum column limit per page is ' + SERVER_SIDE_COLUMN_LIMIT);
+                }
+
+                return this.resultFactory(
+                    executionResponse.links.executionResult,
+                    [offsetRows, offsetColumns],
```
Comment (martinmosko) [inline src/DataLayer/dataSources/DataSource.ts]: are you going to always request with two dimensional data? 

this looks like a dangerous idea

File: src/DataLayer/dataSources/tests/DataSource.spec.ts [hunk 8]
```diff
@@ -32,8 +36,140 @@ describe('DataSource', () => {
...
+
+            expect(responseFactory).toBeCalledWith(resultSpec);
+            expect(resultFactory).toBeCalledWith('url1', [0, 0], [SERVER_SIDE_ROW_LIMIT, SERVER_SIDE_COLUMN_LIMIT]);
+            expect(responses).toEqual({
+                executionResponse: { links: { executionResult: 'url1' } },
+                executionResult: { data: [] }
+            });
+        });
+
+        it('should cache first response', async () => {
```
Comment (martinmosko) [inline src/DataLayer/dataSources/tests/DataSource.spec.ts]: Only testing two dimensional request, what about 1-dim ?

File: src/DataLayer/dataSources/tests/DataSource.spec.ts [hunk 9]
```diff
@@ -32,8 +36,140 @@ describe('DataSource', () => {
...
+                [SERVER_SIDE_ROW_LIMIT, SERVER_SIDE_COLUMN_LIMIT],
+                [10, 10]
+            ]);
+            expect(responses2).toEqual({
+                executionResponse: { links: { executionResult: 'url1' } },
+                executionResult: { data: [[]] }
+            });
+        });
+
+        it('should work with different resultSpec', async () => {
```
Comment (martinmosko) [inline src/DataLayer/dataSources/tests/DataSource.spec.ts]: what does it mean `work with different resultSpec` ?

File: src/DataLayer/dataSources/tests/DataSource.spec.ts [hunk 10]
```diff
@@ -32,8 +36,140 @@ describe('DataSource', () => {
...
+                executionResult: { data: [[], []] }
+            });
+        });
+
+        it('should limit page size by server limit', async () => {
+            const consoleWarnSpy = jest.spyOn(global.console, 'warn');
+            consoleWarnSpy.mockImplementation(jest.fn());
+
+            const { responseFactory, resultFactory, dataSource } = createDataSource();
+            const overLimit = [SERVER_SIDE_ROW_LIMIT + 2, SERVER_SIDE_COLUMN_LIMIT + 2];
```
Comment (martinmosko) [inline src/DataLayer/dataSources/tests/DataSource.spec.ts]: this does not look like testing edge value, please consider that

File: src/execution/execute-afm.ts [hunk 11]
```diff
@@ -44,14 +44,23 @@ export class ExecuteAfmModule {
...
         return this.xhr.post(`/gdc/app/projects/${projectId}/executeAfm`, { body: JSON.stringify(execution) })
             .then(apiResponse => apiResponse.getData())
             .then(unwrapExecutionResponse);
     }
 
+    public fetchExecutionResult(executionResultUri: string, offset: number[], limit: number[])
+        : Promise<Execution.IExecutionResult> {
+        const uri = replaceLimitAndOffsetInUri(executionResultUri, limit, offset);
+
+        return this.xhr.ajax(uri, { method: 'GET' })
```
Comment (martinmosko) [inline src/execution/execute-afm.ts]: `this.xhr.get(...)` less words :)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/36/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/31/ : FAILURE in 4m 00s

File: src/DataLayer/dataSources/tests/DataSource.spec.ts (hunk 9 above)
Comment (tomasreichmann) [reply src/DataLayer/dataSources/tests/DataSource.spec.ts]: That it should be able to handle two different result spec requests

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/37/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/32/ : FAILURE in 1m 43s

**Comment by tomasreichmann** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/38/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/33/ : FAILURE in 2m 56s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/39/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/34/ : FAILURE in 1m 13s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/40/ : SUCCESS in 13s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/35/ : FAILURE in 2m 40s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/41/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/36/ : SUCCESS in 1m 04s

File: src/DataLayer/dataSources/tests/DataSource.spec.ts (hunk 9 above)
Comment (martinmosko) [reply src/DataLayer/dataSources/tests/DataSource.spec.ts]: `should be able to handle two different result spec requests` this is better message than `should work with different resultSpec` :) could you please update it ? :)

File: src/DataLayer/utils/DummyDataSource.ts (hunk 1 above)
Comment (tomasreichmann) [reply src/DataLayer/utils/DummyDataSource.ts]: My IDE still has a problem with this. So I leave the comment in place

File: src/DataLayer/dataSources/DataSource.ts (hunk 6 above)
Comment (martinmosko) [reply src/DataLayer/dataSources/DataSource.ts]: Due to discussion with @tomasreichmann it will be quite difficult to change it and keep backward compatibility

File: src/DataLayer/dataSources/tests/DataSource.spec.ts (hunk 8 above)
Comment (martinmosko) [reply src/DataLayer/dataSources/tests/DataSource.spec.ts]: Test added `it('should work with a single dimension', async () => {`

File: src/DataLayer/dataSources/DataSource.ts (hunk 5 above)
Comment (martinmosko) [reply src/DataLayer/dataSources/DataSource.ts]: No changes right now as discussed with @tomasreichmann

File: src/execution.ts [hunk 12]
```diff
@@ -14,7 +13,21 @@ import { MetadataModule } from './metadata';
  *
  */
 export class ExecutionModule {
-    constructor(private xhr: XhrModule, private md: MetadataModule) {}
+    public readonly executeAfm: ExecuteAfmModule['executeAfm'];
+    public getExecutionResponse: ExecuteAfmModule['getExecutionResponse'];
+    public fetchExecutionResult: ExecuteAfmModule['fetchExecutionResult'];
```
Comment (martinmosko) [inline src/execution.ts]: missing readonly

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/50/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/46/ : SUCCESS in 53s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/51/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/47/ : SUCCESS in 46s

**Comment by martinmosko** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-merge-compliance-check/12/ : SUCCESS in 14s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/48/ : SUCCESS in 59s
